src/rpc/blockchain.cpp

```diff
@@ -1490,14 +1490,73 @@ UniValue MempoolInfoToJSON(const CTxMemPool& pool)
     ret.pushKV("mempoolminfee", ValueFromAmount(std::max(pool.GetMinFee(maxmempool), ::minRelayTxFee).GetFeePerK()));
     ret.pushKV("minrelaytxfee", ValueFromAmount(::minRelayTxFee.GetFeePerK()));
 
+    if (with_fee_histogram) {
+        /* TODO: define log scale formular for dynamically creating the
+         * feelimits but with the property of not constantly changing
+         * (and thus screw up client implementations) */
+        static const std::vector<CAmount> feelimits{1, 2, 3, 4, 5, 6, 7, 8, 10,
+            12, 14, 17, 20, 25, 30, 40, 50, 60, 70, 80, 100,
+            120, 140, 170, 200, 250, 300, 400, 500, 600, 700, 800, 1000,
+            1200, 1400, 1700, 2000, 2500, 3000, 4000, 5000, 6000, 7000, 8000, 10000};
+
+        /* keep histogram per...
+         * ... cumulated tx sizes
+         * ... txns (count)
+         * ... cumulated fees */
+        std::vector<uint64_t> sizes(feelimits.size(), 0);
+        std::vector<uint64_t> count(feelimits.size(), 0);
+        std::vector<uint64_t> fees(feelimits.size(), 0);
+
+        for (const CTxMemPoolEntry& e : pool.mapTx) {
+            int size = (int)e.GetTxSize();
+            CAmount fee = e.GetFee();
+            uint64_t asize = e.GetSizeWithAncestors();
+            CAmount afees = e.GetModFeesWithAncestors();
+            uint64_t dsize = e.GetSizeWithDescendants();
+            CAmount dfees = e.GetModFeesWithDescendants();
+
+            CAmount fpb = fee / size; //fee per byte
+            CAmount afpb = afees / asize; //fee per byte including ancestors
+            CAmount dfpb = dfees / dsize; //fee per byte including descendants
+            CAmount tfpb = (afees + dfees - fee) / (asize + dsize - size);
+            CAmount feeperbyte = std::max(std::min(dfpb, tfpb), std::min(fpb, afpb));
+
+            // distribute feerates into feelimits
+            for (size_t i = 0; i < feelimits.size(); i++) {
+                if (feeperbyte >= feelimits[i] && (i == feelimits.size() - 1 || feeperbyte < feelimits[i + 1])) {
+                    sizes[i] += size;
+                    count[i]++;
+                    fees[i] += fee;
+                    break;
+                }
+            }
+        }
+        CAmount total_fees = 0; //track total amount of available fees in mempool
+        UniValue info(UniValue::VOBJ);
+        for (size_t i = 0; i < feelimits.size(); i++) {
+            UniValue info_sub(UniValue::VOBJ);
+            info_sub.pushKV("sizes", sizes[i]);
```

Comment:
what is sizes? I've noticed that when adding up all the sizes when the mempool is full, that the number doesn't stay fixed as I would have expected when the mempool is full - so it's not the number of bytes used in memory to store the tx - so, what is it?

Reply:
So the histogram is based on fee rates intervals. The histogram is modeled using three vectors:

```cpp
        std::vector<uint64_t> sizes(feelimits.size(), 0);
        std::vector<uint64_t> count(feelimits.size(), 0);
        std::vector<uint64_t> fees(feelimits.size(), 0);
```

where `sizes[0]` represents cumulative size of txs belonging to the first fee rate interval `[1, 2)`, `sizes[1]` represents  cumulative size of txs belonging to the second fee rate interval `[2, 3)`, etc.

[Line 1527](https://github.com/bitcoin/bitcoin/pull/15836/files#diff-decae4be02fb8a47ab4557fe74a9cb853bdfa3ec0fa1b515c0a1e5de91f4ad0bR1527) shows that we add `size` to `sizes[i]` where `size` is defined as `int size = (int)e.GetTxSize();` which is defined as:

```cpp
size_t CTxMemPoolEntry::GetTxSize() const
{
    return GetVirtualTransactionSize(nTxWeight, sigOpCost);
}
```

What is virtual size of a transaction? This is explained here: https://bitcoin.stackexchange.com/questions/92689/how-is-the-size-of-a-bitcoin-transaction-calculated.

HTH!

promo: You may have a look at #21422 too. :)